core-rust/node-common/src/java/utils.rs

```diff
@@ -97,15 +100,60 @@ pub fn jni_sbor_coded_call<Args: ScryptoDecode, Response: ScryptoEncode>(
     jni_sbor_coded_fallible_call(env, encoded_request, |args| Ok(method(args)))
 }
 
+/// A convenience method for a `jni_call()` which uses SBOR codec for its argument and its
+/// potentially-erroneous result.
 pub fn jni_sbor_coded_fallible_call<Args: ScryptoDecode, Response: ScryptoEncode>(
     env: &JNIEnv,
     encoded_request: jbyteArray,
     method: impl FnOnce(Args) -> JavaResult<Response>,
 ) -> jbyteArray {
-    let result = jni_jbytearray_to_vector(env, encoded_request)
-        .and_then(|bytes| Args::from_java(&bytes))
-        .and_then(method);
-    jni_slice_to_jbytearray(env, &result.to_java().unwrap())
+    jni_call(env, || {
+        let result = jni_jbytearray_to_vector(env, encoded_request)
+            .and_then(|bytes| Args::from_java(&bytes))
+            .and_then(method);
+        jni_slice_to_jbytearray(env, &result.to_java().unwrap())
+    })
+    .unwrap_or_else(std::ptr::null_mut)
+}
+
+/// Executes the given function in a way that is panic-safe on a JNI-originating stack.
+/// This is achieved by intercepting any panic (i.e. `catch_unwind()`) and throwing a Java-side
+/// `RustPanicException` instead.
+///
+/// *This is a mandatory wrapper for all JNI calls.*
+/// Every top-level JNI method MUST immediately enter this template method (i.e. before executing
+/// any logic that may panic) - otherwise, unwinding a stack through the JNI boundary results in an
+/// undefined behavior.
+/// It is fine to enter this method indirectly, e.g. via one of convenience variants defined above.
+///
+/// Note on the "abrupt return" from this method:
+/// In case a Java exception is thrown, this function will return `None`, signalling the caller that
+/// there is no valid result of the call, and that it should immediately return from the JNI method,
+/// allowing the Java-side JNI infra to continue with the exception.
+/// For syntactic reasons, the caller may need to return some value back to the JNI - it is then
+/// fine to return virtually any value of a valid type (e.g. a `null_mut()` pointer, which denotes
+/// a Java `null`), since it will be ignored by the Java-side JNI infra anyway.
+pub fn jni_call<R>(env: &JNIEnv, callable: impl FnOnce() -> R) -> Option<R> {
+    let result = panic::catch_unwind(AssertUnwindSafe(callable)).map_err(|panic_payload| {
+        if let Some(string) = panic_payload.downcast_ref::<String>() {
+            string.clone()
+        } else if let Some(str_ref) = panic_payload.downcast_ref::<&'static str>() {
+            str_ref.to_string()
+        } else {
+            format!("[non-string panic payload: {:?}]", panic_payload)
+        }
+    });
+    match result {
+        Ok(return_value) => Some(return_value),
+        Err(panic_message) => {
+            let throw_result =
+                env.throw_new("com/radixdlt/exceptions/RustPanicException", panic_message);
```

Comment:
How exactly does this work? Does this pass back to the caller? Or only raise the exception when this method returns? (hence the return of a null pointer)?

Reply:
I was imagining we'd return a RustPanicException using a StateManagerError on JavaResult, but this works too / maybe is nicer :D

Reply:
> Or only raise the exception when this method returns? (hence the return of a null pointer)?

Yes, I tried to explain it in the `Note on the "abrupt return" from this method` rustdoc.
Basically - rust has no throws, so it must give the control back to Java in a regular "return" way.

> we'd return a RustPanicException using a StateManagerError on JavaResult

I thought about this for a moment, but I decided it is more reliable to just throw the exception on panic (i.e. not relying on Java side to correctly interpret certain JavaResults). Especially since there are currently some cases of JNI calls where we do _not_ use the `Natives.builder()` infra (i.e. where there is no handling for `JavaResult`), e.g. [here](https://github.com/radixdlt/babylon-node/blob/f9a780ad9f53dad034644f4846e50da7b5e395f1/core-rust-bridge/src/main/java/com/radixdlt/statemanager/StateManager.java#L95).